### PR TITLE
fix(mcp): gate TDD test execution

### LINKED ISSUE #3368: Harden TDD MCP test execution against shell-command input
# Harden TDD MCP test execution against shell-command input

## Summary

The TDD MCP server exposes `check_red_green_cycle.testCommand` as free-form command text and passes it to `child_process.execSync()` as a shell command string. The related compliance-analysis path also executes `npm test --silent` when invoked, without an explicit approval or dry-run boundary.

## Security impact

A prompt-injected or untrusted MCP client that can call the TDD server can cross from tool input into local child-process execution under the MCP server process. In developer or CI contexts, the process may have access to source files, local credentials, or environment variables. Actual blast radius depends on how the MCP server is deployed and what privileges its process has.

## Evidence

- `repo/src/mcp-server/schemas.ts:243-245` accepts `testCommand` as an unconstrained string with default `npm test`.
- `repo/src/mcp-server/tdd-server.ts:114-128` exposes `check_red_green_cycle` and its `testCommand` input.
- `repo/src/mcp-server/tdd-server.ts:152-167` dispatches the MCP tool call to `checkRedGreenCycle()`.
- `repo/src/mcp-server/tdd-server.ts:289-294` interpolates the parsed command text into `execSync()`.
- `repo/src/mcp-server/tdd-server.ts:177-213` runs `npm test --silent` from a compliance-analysis tool without a dry-run/approval gate.

No live MCP invocation, test command execution, dependency installation, network request, or exploit attempt was performed during the audit.

## Recommended remediation

- Use `execFile` or `spawn` with a fixed executable and argv array.
- Allow only configured test runners and known-safe argument sets.
- Require explicit operator approval for test execution, and make dry-run/plan output the default where feasible.
- Constrain execution cwd to an approved workspace.
- Keep analysis-only MCP tools from implicitly executing package scripts.

## Regression test idea

Mock the child-process API and assert that invalid/free-form command text is rejected before execution. Assert that valid runner invocations are passed as executable plus argv, not shell strings. Add a test that `analyze_tdd_compliance` does not execute tests unless an explicit approval/dry-run policy is satisfied.

### PR DESCRIPTION
**Pull request by ootakazuhiko**
## Summary

Closes #3368.

This PR hardens the TDD MCP server's test-execution boundary:

- replaces free-form shell command execution with an allowlisted executable/argv policy;
- makes TDD MCP test execution dry-run by default;
- requires the explicit approval token `approved-tdd-test-execution` plus `dryRun=false` before tests run;
- uses `execFileSync` with `shell: false` for approved runs;
- prevents `analyze_tdd_compliance` from implicitly running package scripts unless the same approval policy is satisfied;
- constrains the analysis/execution cwd to the approved workspace root.

## Acceptance

- [x] Free-form `testCommand` strings such as shell injection payloads are rejected by schema validation.
- [x] Approved commands resolve to executable plus argv, not shell strings.
- [x] `check_red_green_cycle` does not execute by default and reports the approved command plan.
- [x] `analyze_tdd_compliance` does not implicitly run tests unless execution is explicitly approved.
- [x] Approved execution uses `shell: false`.
- [x] Workspace path escapes are rejected before test execution cwd is used.

## Verification

Local verification run on this branch:

```bash
pnpm install --frozen-lockfile
pnpm -s exec vitest run tests/unit/mcp-server/tdd-execution-policy.test.ts --reporter dot
pnpm -s exec eslint --quiet src/mcp-server/tdd-server.ts src/mcp-server/tdd-execution-policy.ts src/mcp-server/schemas.ts tests/unit/mcp-server/tdd-execution-policy.test.ts
pnpm -s run types:check
pnpm -s run build
pnpm -s run check:schemas
pnpm -s run check:doc-consistency
git diff --check
```

## Rollback

Revert this PR to restore the previous TDD MCP behavior. Rollback reopens #3368 risk because MCP input would again be able to reach shell command execution without an approval/dry-run boundary.

### CODE REVIEW

File: src/mcp-server/tdd-server.ts [hunk 1]
```diff
@@ -338,6 +400,26 @@ class TDDGuardServer {
     return stdout || stderr;
   }
 
+  private resolveApprovedWorkspacePath(inputPath: string): string {
+    if (inputPath.includes('\0') || inputPath.includes('\\')) {
+      throw new Error('TDD analysis path must not contain NUL bytes or backslashes');
+    }
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply src/mcp-server/tdd-server.ts:415]: Addressed in `82a20b20`: removed the backslash rejection from `resolveApprovedWorkspacePath()`. The path guard now rejects NUL bytes and relies on `path.resolve` / `realpathSync` / `path.relative` containment so Windows paths remain supported.

File: src/mcp-server/tdd-server.ts [hunk 2]
```diff
@@ -287,10 +333,26 @@ class TDDGuardServer {
...
+            `Approved executable/argv: ${formatTDDTestCommand(resolvedCommand)}.`,
+            `Set dryRun=false and testExecutionApproval='approved-tdd-test-execution' to execute without shell interpretation.`,
+          ].join('\n'),
+        }],
+      };
+    }
     
     try {
-      execSync(`${testCommand} --silent`, { encoding: 'utf8', stdio: 'pipe' });
+      runSafeTDDTestCommand(testCommand);
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply src/mcp-server/tdd-server.ts:365]: Addressed in `82a20b20`: approved `check_red_green_cycle` executions now pass `cwd: this.resolveApprovedWorkspacePath('.')`. Added a regression test with a fake `npm` executable to assert the command runs from the approved workspace root.

File: src/mcp-server/tdd-server.ts [hunk 3]
```diff
@@ -208,31 +244,41 @@ class TDDGuardServer {
...
+        `Test execution skipped (${executionApproval.reason}). Approved command would be: ${formatTDDTestCommand(resolvedCommand)}`;
     }
 
     // Generate recommendations
     analysis.recommendations = this.generateRecommendations(analysis.violations, phase);
+    if (skippedTestExecutionRecommendation) {
+      analysis.recommendations.push(skippedTestExecutionRecommendation);
+    }
     analysis.nextAction = this.determineNextAction(analysis.violations, phase);
     analysis.canProceed = analysis.violations.filter(v => v.severity === 'error').length === 0;
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply src/mcp-server/tdd-server.ts:292]: Addressed in `82a20b20`: skipped execution in phases that require RED/GREEN evidence (`3-tests` and `4-code`) now records an error-level `test_execution_skipped` violation, making the report non-proceedable. Added a `4-code` regression test covering the default dry-run path.

File: src/mcp-server/tdd-execution-policy.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    cwd?: string;
+    executor?: TDDTestExecutor;
+  } = {}
+): string | Buffer {
+  const resolved = resolveSafeTDDTestCommand(testCommand);
+  const executor = options.executor ?? execFileSync;
+  const execOptions: { encoding: 'utf8'; stdio: 'pipe'; shell: false; cwd?: string } = {
+    encoding: 'utf8',
+    stdio: 'pipe',
+    shell: false,
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply src/mcp-server/tdd-execution-policy.ts:82]: Addressed in `82a20b20`: Windows approved commands now resolve to a fixed `cmd.exe /d /s /c <manager>.cmd test --silent` argv while retaining `shell: false` and the allowlisted command surface. Added direct Windows-resolution and execution-policy tests.
